R/cache-memory.R

```diff
@@ -0,0 +1,344 @@
+      invisible(self)
+    },
+
+    exists = function(key) {
+      validate_key(key)
+      exists(key, envir = private$cache, inherits = FALSE)
+    },
+
+    keys = function() {
+      ls(private$cache, sorted = FALSE)  # Faster with sorted=FALSE
+    },
+
+    remove = function(key) {
+      private$log(paste0('remove: key "', key, '"'))
+      validate_key(key)
+      rm(list = key, envir = private$cache)
+      invisible(self)
+    },
+
+    reset = function() {
+      private$log(paste0('reset'))
+      rm(list = self$keys(), envir = private$cache)
+      invisible(self)
+    },
+
+    prune = function() {
+      private$log(paste0('prune'))
+      info <- private$object_info()
+
+      # 1. Remove any objects where the age exceeds max age.
+      time <- as.numeric(Sys.time())
+      timediff <- time - info$mtime
+      rm_idx <- timediff > private$max_age
+      if (any(rm_idx)) {
+        private$log(paste0("prune max_age: Removing ", paste(info$key[rm_idx], collapse = ", ")))
+      }
+
+      # Remove items from cache
+      rm(list = info$key[rm_idx], envir = private$cache)
+      info <- info[!rm_idx, ]
+
+      # Sort objects by priority, according to eviction policy.
+      if (private$evict == "lru") {
+        info <- info[order(info[["atime"]], decreasing = TRUE), ]
+      } else if (private$evict == "fifo") {
+        info <- info[order(info[["mtime"]], decreasing = TRUE), ]
+      } else {
+        stop('Unknown eviction policy "', private$evict, '"')
+      }
+
+      # 2. Remove objects if there are too many.
+      if (nrow(info) > private$max_n) {
+        rm_idx <- seq_len(nrow(info)) > private$max_n
+        if (any(rm_idx)) {
+          private$log(paste0("prune max_n: Removing ", paste(info$key[rm_idx], collapse = ", ")))
+        }
+        rm(list = info$key[rm_idx], envir = private$cache)
+        info <- info[!rm_idx, ]
+      }
+
+      # 3. Remove objects if cache is too large.
+      if (sum(info$size) > private$max_size) {
+        cum_size <- cumsum(info$size)
+        rm_idx <- cum_size > private$max_size
+        if (any(rm_idx)) {
+          private$log(paste0("prune max_size: Removing ", paste(info$key[rm_idx], collapse = ", ")))
+        }
+        rm(list = info$key[rm_idx], envir = private$cache)
+        info <- info[!rm_idx, ]
+    }
+      invisible(self)
+    },
+
+    size = function() {
+      # TODO: Validate this against metadata?
+      length(self$key())
+    },
+
+    destroy = function() {
+      if (is.null(private$cache)) {
+        return(invisible)
+      }
+      private$log(paste0("destroy"))
+
+      private$cache <- NULL
+    },
+
+    is_destroyed = function() {
+      is.null(private$cache)
+    }
+  ),
+
+  private = list(
+    cache = NULL,
+    meta = NULL,     # Metadata used for pruning
+    max_age = NULL,
+    max_size = NULL,
+    max_n = NULL,
+    evict = NULL,
+    missing = NULL,
+    exec_missing = NULL,
+    logfile = NULL,
+
+    object_info = function() {
```

Comment:
General comment: Consider the common code paths (getting, or setting without a prune actually removing items) and making them fast, by e.g. tracking state in cheap-to-consult fields.